### ISSUE TITLE
MvxAutoComplete quits working when you backspace until the TextView is empty.

### ISSUE THREAD

**Issue by JimWilcox3** [MvvmCross/MvvmCross#945]
I have created a repo to demonstrate this bug.  I am using an MvxAutoComplete in a Data Template for an MvxItemList.  If you run the demo below, There are 5 lines with data.  On the first line, start backspacing.  As long as you don't erase the "B" at the beginning it works fine.  Once the field is empty, it quits working.  One thing I have noticed is that if there is only on character in field, if you look at the trace, it is looking to match an empty string rather than the one character you might expect.  If you start typing on any of the blank lines, no autocomplete suggestions are ever made..

For further testing, I removed the binding for the Text property of the AutoComplete field.  If you do this then if you start typing on a blank line, it will work.  However, if you backspace enough to erase the first character, it quits working.

https://github.com/JimWilcox3/MvxAutoCompleteTest/tree/master/MvxAutoCompleteTest

Please let me know if this is a real bug or if I did something wrong or if there is a work around.

Thanks,
Jim

**Comment by munkii** [comment]
That control is a PITA but I believe that is mainly because of the Android control it is wrapping rather than  the MVVM Cross wrapper. I have it working across my Android app and posted an answer to my own question on SO, https://stackoverflow.com/questions/26543931/partialtextchanged-stops-firing-on-mvxautocompletetextview-after-item-selection. Here's another issue I had with THAT control, https://stackoverflow.com/questions/26257083/mvxautocompletetextview-setting-text-property-to-selectedobject-tostring. If these don't help I will try to get some time later to have a look at your github repo

**Comment by JimWilcox3** [comment]
@munkii Thanks for your reply.  I looked at your examples.  I can say that in my example there isn't a path in the PartialText property that doesn't set the ItemsSource to List<string> (which fires the changed event).  I saw your second example and I am binding to the Text property, but that is solely to capture the resulting input to may database table.  In the example it is a simple field on the ViewModel.  The truth as I mentioned above is that the control works correctly if you don't bind the Text field, but I don't see another way to capture the resulting input.  SelectedObject is there, but the result might not be an exact match to what's in the list.

**Comment by JimWilcox3** [comment]
Did some more testing and I think I know where the problem lies.  I added a trace to my property bound to PartialText which is called CurrentTextHint

```
        private string _currentTextHint;
        public string CurrentTextHint
        {
            get
            { return _currentTextHint; }
            set
            {
                MvxTrace.Trace("Partial Text Value Sent {0}", value);
                _currentTextHint = value;

                if (_currentTextHint.Trim().Length < 2)
                {
                    SetSuggestionsEmpty();
                    return;
                }

                var list = AutoCompleteList.Where(i => (i ?? "").ToUpper().Contains(_currentTextHint.ToUpper()));
                if (list.Count() > 0)
                {
                    AutoCompleteSuggestions = list.ToList();
                }
                else
                {
                    SetSuggestionsEmpty();
                }
            }
        }
```

If you consider the code in MvxFilteringAdapter for SetConstraintAndWaitForDataChange, you can see there are several traces in there as well:

```
        private int SetConstraintAndWaitForDataChange(string newConstraint)
        {
            MvxTrace.Trace("Wait starting for {0}", newConstraint);
            _dataChangedEvent.Reset();
            this.PartialText = newConstraint;
            _dataChangedEvent.WaitOne();
            MvxTrace.Trace("Wait finished with {1} items for {0}", newConstraint, Count);
            return Count;
        }
```
Consider the trace when I am backspacing through the word Bed:

[0:] 
mvx:Diagnostic: 29.71 Wait starting for Bed
[0:] mvx:Diagnostic: 29.71 Wait starting for Bed
03-25 15:58:02.232 I/mono-stdout( 3067): mvx:Diagnostic: 29.71 Wait starting for Bed
[0:] 
mvx:Diagnostic: 29.76 Partial Text Value Sent Bed
03-25 15:58:02.302 I/mono-stdout( 3067): mvx:Diagnostic: 29.76 Partial Text Value Sent Bed
[0:] mvx:Diagnostic: 29.76 Partial Text Value Sent Bed
[0:] 
mvx:Diagnostic: 29.88 Wait finished with 4 items for Bed
[0:] mvx:Diagnostic: 29.88 Wait finished with 4 items for Bed
03-25 15:58:02.412 I/mono-stdout( 3067): mvx:Diagnostic: 29.88 Wait finished with 4 items for Bed
Thread finished:  #10
The thread 'Unknown' (0xa) has exited with code 0 (0x0).
Thread started:  #11
[0:] 
mvx:Diagnostic: 38.32 Wait starting for Be
03-25 15:58:10.882 I/mono-stdout( 3067): mvx:Diagnostic: 38.32 Wait starting for Be
[0:] mvx:Diagnostic: 38.32 Wait starting for Be
[0:] 
mvx:Diagnostic: 38.45 Partial Text Value Sent Be
[0:] mvx:Diagnostic: 38.45 Partial Text Value Sent Be
03-25 15:58:10.962 I/mono-stdout( 3067): mvx:Diagnostic: 38.45 Partial Text Value Sent Be
[0:] 
mvx:Diagnostic: 38.47 Wait finished with 4 items for Be
[0:] mvx:Diagnostic: 38.47 Wait finished with 4 items for Be
03-25 15:58:10.992 I/mono-stdout( 3067): mvx:Diagnostic: 38.47 Wait finished with 4 items for Be
[0:] 
mvx:Diagnostic: 41.21 Wait starting for 
[0:] mvx:Diagnostic: 41.21 Wait starting for 
03-25 15:58:14.732 I/mono-stdout( 3067): mvx:Diagnostic: 41.21 Wait starting for 
[0:] 
mvx:Diagnostic: 42.24 Partial Text Value Sent 
[0:] mvx:Diagnostic: 42.24 Partial Text Value Sent 
03-25 15:58:14.762 I/mono-stdout( 3067): mvx:Diagnostic: 42.24 Partial Text Value Sent 
[0:] 
mvx:Diagnostic: 42.31 Wait finished with 0 items for 
[0:] mvx:Diagnostic: 42.31 Wait finished with 0 items for 
03-25 15:58:14.872 I/mono-stdout( 3067): mvx:Diagnostic: 42.31 Wait finished with 0 items for 
[0:] 
mvx:Diagnostic: 45.25 Wait starting for 
[0:] mvx:Diagnostic: 45.25 Wait starting for 
03-25 15:58:17.762 I/mono-stdout( 3067): mvx:Diagnostic: 45.25 Wait starting for 

When I backspace so that only the letter B is showing, SetConstraintAndWaitForDataChange sets the constraint to an empty string.  This might be by design, but it does return with a "Wait finished".  When I backspace one more time, It gets the "Wait starting" trace but never gets the "Partial Text" trace which I had added in the property.  This tells me that in the SetConstraintAndWaitForDataChange that either

```
_dataChangedEvent.Reset();
```
or 
```
this.PartialText = newConstraint;
```
is throwing an error or failing to happen for some reason.  Whatever is happening, it is never setting the value for the CurrentTextHint property.  I hope someone will be willing to check into this and fix it.  Any diagnosis beyond this is above me.

**Comment by JimWilcox3** [comment]
@slodge @Cheesebaron @munkii I found an easy workaround for this.  Are any of you guys interested  in what I found?  I can do the long version, short version, or you can just close this out.  Just let me know.

**Comment by Dexyon** [comment]
Didn't you have a github project to show the problem. Maybe you can update the project and close this issue.

**Comment by JimWilcox3** [comment]
@Dexyon I did update my repo so that it works correctly.  

https://github.com/JimWilcox3/MvxAutoCompleteTest

The reason I asked if an explanation was requested is that upon my initial googling of this issue, it appears that many people have run into this issue in some form or another.  I downloaded the MvvmCross source and worked through what was causing the problem.  The fix is actually at the user code level and not something in MvvmCross.  The problem actually stems from the actual Android Widget that is being wrapped.  The fix is pretty simple, but the explanation of the fix is pretty in depth so I didn't want to spend a lot of time explaining it if no one was going to read it.

**Comment by JimWilcox3** [comment]
I am going to go ahead and explain what I did to work around this issue and close it.  Hopefully this information will help the developers or someone else in the future.

There are several dynamics which must be understood when using the MvxAutoComplete.

First, MvxAutoComplete relies on MvxFilteringAdapter.  When text is typed into the control, MvxFitleringAdapter sends the text back to the property bound to the PartialText Property of the MvxAutoComplete.  It sets a manual event callback that requires a PropertyChanged event to be fired on the property bound to the ItemsSource property of the control.

Whatever is responsible for setting the PartialText bound property first does a get on that property.  If the get returns the same thing that the control wants to pass in the setter, it doesn't call the setter.  The setter for the PartialText bound property is what is responsible for creating a suggestion list and setting the value in the ItemsSource and raising the event to notify that it has changed.  If the setter is never called, then the MvxFilteringAdapter gets stuck waiting for the event to happen.

When text is being typed into the control, if the length of the string is less that 2, then it returns an empty string.  If the PartialText property is an empty string and it passes it a new empty string, the setter is not called.  An important note is that if you have the Text property bound to your ViewModel, then it calls the setter for the PartialText during the binding process, so the current PartialText is already set to an empty string before you even interact with it.  Also, if there is a value already in the Text bound field and you start backspacing, when it gets down to one character in length, then the control passes an empty string to the setter for PartialText so when you backspace one more time to blank out the field, it has gotten two empty strings in a row so again it doesn't call the setter and hangs up the MvxFitleringAdapter.

So what to do about it?  In the setter for the PartialText, if "value" is an empty string, just set the field to null.  Pulling my example from one that @slodge wrote a while back, I created a property called CurrentTextHint which I bound to the PartialText property of the control.  Here is how I implemented it.

```
        private string _currentTextHint;
        public string CurrentTextHint
        {
            get
            { return _currentTextHint; }
            set
            {
                MvxTrace.Trace("Partial Text Value Sent {0}", value);
                //Setting _currentTextHint to null if an empty string gets passed here
                //is extremely important.
                if (value == "")
                {
                    _currentTextHint = null;
                    SetSuggestionsEmpty();
                    return;
                }
                else
                {
                    _currentTextHint = value;
                }

                if (_currentTextHint.Trim().Length < 2)
                {
                    SetSuggestionsEmpty();
                    return;
                }

                var list = AutoCompleteList.Where(i => (i ?? "").ToUpper().Contains(_currentTextHint.ToUpper()));
                if (list.Count() > 0)
                {
                    AutoCompleteSuggestions = list.ToList();
                }
                else
                {
                    SetSuggestionsEmpty();
                }
            }
        }
```

The MvxAutoComplete never tries to pass a null to PartialText so setting it to null when you get an empty string ensures that the value will be different from what is already there thus the setter will always be called when the text changes in the control.

I have a working version at:

https://github.com/JimWilcox3/MvxAutoCompleteTest

I hope this is helpful.

Jim
